### ISSUE TITLE
session-manager crashed cause of async issue(already have root cause analysis need advise for solution)

### ISSUE THREAD

**Issue by w2angrong** [anbox/anbox#1474]
1. Please check that no similar bug is already reported. Have a look on the list of open bugs at https://github.com/anbox/anbox/issues
2. Make sure you are running the latest version of Anbox before reporting an issue. Update snap to latest: `snap refresh --devmode --edge anbox`
3. Make sure you have debug logs enabled:
`sudo snap set anbox debug.enable=true`
4. Reproduce the error while debug logs enabled.
5. Run the anbox logs collection utility and attach the tar file.
`sudo /snap/bin/anbox.collect-bug-info `

6. ** Please paste the result of `anbox system-info` below:**
<details>
<summary>anbox system-info output</summary>

```
[please paste printout of `anbox system-info` here]
```
</details>

**Please describe your problem:**

We found there is an asynchronous issue in anbox design. Basically, in anbox the main event loop in anbox/platform/sdl/Platform::process_events() will handle almost everything for window events, including window focusIn and focusOut. But it will create/destroy sdl_window in MultiWindowManager::apply_window_state_update(), which actually, in another thread designed for communicating with Android. There are no effective measures to avoid this two threads using the same window data at the same time, causing race condition.
Assuming, I pressed the close button of one window, anbox will tell this event to Android, after Android get the message and destroy the window, MultiWindowManager::apply_window_state_update() will get a destroyed window message back and try to destroy the sdl_window and relevant data. But during Android handle the message, anything can happened, like before we destroy the sdl_window, I pressed another window and cause a focusOut event of current window, then sdl video instance like x11, will try to use current window, but at this moment, we got the sdl_window destroy message from Android and start to deconstruct sdl_window, which actually using by x11, then x11 may will refer a pointer already freed. Then segment fault will happen.
Please find more details in the attached doc.

**What were you expecting?:**

Basically, we almost can confirm this defect is there, and we can reproduce the crash by click close button at the upper right corner frequently and quickly after

open lots of Android applications, but still need anbox experts to check our analysis and it will be my pleasure if you can give us advise about how to handle this issue.

**Additional info:**
[one asynchronous process willcause anbox session manager crash.docx](https://github.com/anbox/anbox/files/4696572/one.asynchronous.process.willcause.anbox.session.manager.crash.docx)

**Comment by w2angrong** [comment]
@necessarily-equal Hello, Sir, thank you for your attention.
Any chance to give some suggestion ?

**Comment by necessarily-equal** [comment]
(I'd like not to close this, commenting)

Le mer 30 sept. 2020 à 6:03, stale[bot] <notifications@github.com> a 
écrit :
